### DISCUSSION TITLE
Building for Android

### DISCUSSION THREAD

**Discussion by mendelct** [koreader/koreader#10972] [General]
The toolchain instructions for Android need to be updated. I believe the correct NDK for Android is r23c. The environment variables listed don't work, for SDK, ANDROID_HOME seems to be correct, and for NDK, ANDROID_NDK_HOME seems to be correct.

**Comment by NiLuJe** [comment]
Which instructions are we talking about?

In my experience, there's no real need for any, `./kodev release -d android` takes care of the heavy lifting. Then again, I use a sane distro, and do have the basic Android tools installed.

**Reply by Frenzie** [reply]
https://github.com/koreader/koreader/blob/master/doc/Building_targets.md I imagine

**Reply by mendelct** [reply]
correct.

**Reply by mendelct** [reply]
It should probably also have under the building section instructions for android x64

**Reply by NiLuJe** [reply]
Right, I'm not entirely sure anybody has tested building with an out-of-tree NDK in pretty much forever, but that might just be me ;o).

**Reply by mendelct** [reply]
I do :) Sometimes my local version get messed up and the easiest way for me to fix it is to just re-clone everything. If I have to have redownload the ndk etc. it can easily double the compile time, if not more.

**Reply by benoit-pierre** [reply]
How do you get in such a state that re-cloning is necessary? Have you tried `make dist-clean`?

**Reply by mendelct** [reply]
I am far from a programmer, I don't know what `make dist-clean` is (: I basically cobbled together some patches that work for me and run them before compiling. Sometimes after an update I get a makefile error, not knowing how to deal with it, I just delete and re-clone.
The patches I apply can't be user patches e.g. I want the time in cre not to have a leading zero I need to add `if ( *str == '0' ) { memmove(str, str+1, strlen(str)); }` to lvdocview.cpp

**Comment by mendelct** [comment]
This just happened to me, after running `make dist-clean`, don't begin to know what to do. The easiest way for me is to just re-clone.
```
==== Building LuaJIT 2.1 (amalgamation) ====

make -C src amalg

make[4]: Entering directory '/koreader/platform/android/luajit-launcher/jni/luajit/build/arm64-v8a/src'

make all "LJCORE_O=ljamalg.o"

make[5]: Entering directory '/koreader/platform/android/luajit-launcher/jni/luajit/build/arm64-v8a/src'

make[5]: *** No rule to make target 'luajit.h', needed by 'host/buildvm.o'.  Stop.

make[5]: Leaving directory '/koreader/platform/android/luajit-launcher/jni/luajit/build/arm64-v8a/src'

make[4]: *** [Makefile:613: amalg] Error 2

make[4]: Leaving directory '/koreader/platform/android/luajit-launcher/jni/luajit/build/arm64-v8a/src'

make[3]: *** [Makefile:165: amalg] Error 2

make[3]: Leaving directory '/koreader/platform/android/luajit-launcher/jni/luajit/build/arm64-v8a'

make[2]: *** [Makefile:69: build-luajit] Error 2

make[2]: Leaving directory '/koreader/platform/android/luajit-launcher'

make[1]: *** [Makefile:386: androidupdate] Error 2

make[1]: Leaving directory '/koreader'

make: *** [Makefile:554: update] Error 2
```
